43-marabou-stork/technical_specification.md

```diff
@@ -0,0 +1,60 @@
+# Correlation ID Exploration (Marabou Stork)
+**Epic Type:** Implementation Epic
+
+Epic planning and implementation follow the
+[Epic Planning and Marathon SOP](https://docs.ghga-dev.de/main/sops/sop001_epic_planning.html).
+
+**Attention: Please do not put any confidential content here.**
+
+## Scope
+### Outline:
+The aim of this epic is to implement the functionality required to use interservice correlation IDs for tracking associated requests.
+Please see the prototype [here](https://github.com/ghga-de/prototype-correlation-id).
+
+### Included/Required:
+- Generate Correlation ID in API Gateway
+- Correlation ID Utilities in Hexkit
+- Correlation ID in Kafka Event Header
+- Middleware for FastAPI Apps
+- Pilot Repository
+
+### Not Included:
+- Updating services to use correlation IDs
+
+
+## Additional Implementation Details:
+
+### Generate Correlation ID in API Gateway
+The API Gateway will need to be updated with functionality to ensure every request has a correlation ID in the header.
+
+### Correlation ID Utilities in Hexkit
+ContextVars can only be referenced via a single instance, so it is not possible to instantiate a correlation ID ContextVar from one library and set it via another library without importing it. Therefore, `hexkit` should be made a production dependency of `ghga-service-commons`, and correlation ID utilities should be defined within `hexkit`.
+
+The following are needed in `hexkit`:
+- A ContextVar to store the correlation ID.
+- An asynccontextmanager to set the correlation ID ContextVar, yield, and then reset the token.
+- A function to retrieve the value stored in the ContextVar and raises an error if none is found.
+- A function that validates a correlation ID value and raises an error if the value is empty or invalid.
+
+### Correlation ID in Kafka Event Header
+There needs to be a way to supply the correlation ID as a Kafka event header. If the correlation ID is not supplied, one should be generated, but only if a flag has been explicitly set. Otherwise, an error should be raised.
```

Comment:
Do we really need to supply the correlation ID to the publish method? I thought it would use the value from the context var, or create one (depending on a publisher config flag)? In both cases, no input would be needed?

Reply:
No, you're right it doesn't. I'm not sure why I got hung up on that. It should do like you said and grab the value from the ContextVar, then pass that with the _publish_validated() call.

Reply:
I think it was in your notes because we discussed both options for handling cases where we expect no corr ID to be set as context var (e.g. for cron jobs):
1. set a flag to not expect but autogenerate a corr id -> Advantage that we don't expose the generation logic to the core
2. directly provide a custom corr id to the publish method -> Advantage that we are even more flexible that option 1.

Since the advantage of option 1 is not really that big because it is really just `uuid4()` to generate  a new corr ID, I would slightly prefer option 2. However, I would leave the decision to you. Both is absolutely fine.